# Descripción del Proyecto

En este proyecto, compararemos las preferencias musicales de las ciudades de Springfield y Shelbyville. Estudiaremos datos reales para probar tres hipótesis  y compararemos el comportamiento del usuario de esas dos ciudades.

<h1>Tabla de contenidos<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Objetivo" data-toc-modified-id="Objetivo-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Objetivo</a></span></li><li><span><a href="#Etapas" data-toc-modified-id="Etapas-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Etapas</a></span></li><li><span><a href="#Etapa-1.-Descripción-de-los-datos" data-toc-modified-id="Etapa-1.-Descripción-de-los-datos-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>Etapa 1. Descripción de los datos</a></span><ul class="toc-item"><li><span><a href="#Conclusiones-iniciales" data-toc-modified-id="Conclusiones-iniciales-3.1"><span class="toc-item-num">3.1&nbsp;&nbsp;</span>Conclusiones iniciales</a></span></li></ul></li><li><span><a href="#Etapa-2.-Preprocesamiento-de-datos" data-toc-modified-id="Etapa-2.-Preprocesamiento-de-datos-4"><span class="toc-item-num">4&nbsp;&nbsp;</span>Etapa 2. Preprocesamiento de datos</a></span><ul class="toc-item"><li><span><a href="#Estilo-del-encabezado" data-toc-modified-id="Estilo-del-encabezado-4.1"><span class="toc-item-num">4.1&nbsp;&nbsp;</span>Estilo del encabezado</a></span></li><li><span><a href="#Valores-ausentes" data-toc-modified-id="Valores-ausentes-4.2"><span class="toc-item-num">4.2&nbsp;&nbsp;</span>Valores ausentes</a></span></li><li><span><a href="#Duplicados" data-toc-modified-id="Duplicados-4.3"><span class="toc-item-num">4.3&nbsp;&nbsp;</span>Duplicados</a></span></li><li><span><a href="#Conclusiones-" data-toc-modified-id="Conclusiones--4.4"><span class="toc-item-num">4.4&nbsp;&nbsp;</span>Conclusiones <a id="data_preprocessing_conclusions"></a></a></span></li></ul></li><li><span><a href="#Etapa-3.-Prueba-de-hipótesis-" data-toc-modified-id="Etapa-3.-Prueba-de-hipótesis--5"><span class="toc-item-num">5&nbsp;&nbsp;</span>Etapa 3. Prueba de hipótesis <a id="hypotheses"></a></a></span><ul class="toc-item"><li><span><a href="#Hipótesis-1:-comparar-el-comportamiento-del-usuario-en-las-dos-ciudades-" data-toc-modified-id="Hipótesis-1:-comparar-el-comportamiento-del-usuario-en-las-dos-ciudades--5.1"><span class="toc-item-num">5.1&nbsp;&nbsp;</span>Hipótesis 1: comparar el comportamiento del usuario en las dos ciudades <a id="activity"></a></a></span></li><li><span><a href="#Hipótesis-2:-música-al-principio-y-al-final-de-la-semana-" data-toc-modified-id="Hipótesis-2:-música-al-principio-y-al-final-de-la-semana--5.2"><span class="toc-item-num">5.2&nbsp;&nbsp;</span>Hipótesis 2: música al principio y al final de la semana <a id="week"></a></a></span></li><li><span><a href="#Hipótesis-3:-preferencias-de-género-en-Springfield-y-Shelbyville-" data-toc-modified-id="Hipótesis-3:-preferencias-de-género-en-Springfield-y-Shelbyville--5.3"><span class="toc-item-num">5.3&nbsp;&nbsp;</span>Hipótesis 3: preferencias de género en Springfield y Shelbyville <a id="genre"></a></a></span></li></ul></li></ul></div>

## Objetivo  

Hipótesis: 
1. La actividad de los usuarios difiere según el día de la semana y dependiendo de la ciudad. 
2. Los lunes por la mañana, los habitantes de Springfield y Shelbyville escuchan diferentes géneros. Lo mismo ocurre con los viernes por la noche. 
3. Los oyentes de Springfield y Shelbyville tienen preferencias distintas. En Springfield prefieren el pop mientras que en Shelbyville hay más aficionados al rap.

## Etapas 
Los datos del comportamiento del usuario se almacenan en el archivo `/datasets/music_project_en.csv`. No hay ninguna información sobre la calidad de los datos así que deberemos examinarlos antes de probar las hipótesis. 

Evaluaremos la calidad de los datos y veremos si los problemas son significativos.

## Etapa 1. Descripción de los datos

Analisis inicial de datos:

In [1]:
# importamos librerías
import pandas as pd
from IPython.display import display, HTML

Leemos el archivo `music_project_en.csv` de la carpeta `/datasets/` y lo guardamos en la variable `df`:

In [2]:
# leyendo el archivo y almacenándolo en df
df = pd.read_csv('/datasets/music_project_en.csv')

#Miramos las primera 10 filas de nuestro DataFrame
display(df.head(10))

userID                        Track            artist   genre  \
0  FFB692EC            Kamigata To Boots  The Mass Missile    rock   
1  55204538  Delayed Because of Accident  Andreas Rönnberg    rock   
2    20EC38            Funiculì funiculà       Mario Lanza     pop   
3  A3DD03C9        Dragons in the Sunset        Fire + Ice    folk   
4  E2DC1FAE                  Soul People        Space Echo   dance   
5  842029A1                       Chains          Obladaet  rusrap   
6  4CB90AA5                         True      Roman Messer   dance   
7  F03E1C1F             Feeling This Way   Polina Griffith   dance   
8  8FA1D3BE                     L’estate       Julia Dalia  ruspop   
9  E772D5C0                    Pessimist               NaN   dance   

        City        time        Day  
0  Shelbyville  20:28:33  Wednesday  
1  Springfield  14:07:09     Friday  
2  Shelbyville  20:58:07  Wednesday  
3  Shelbyville  08:37:09     Monday  
4  Springfield  08:34:34     Monday  
5  Shelbyville  13:09:41     Friday  
6  Springfield  13:00:07  Wednesday  
7  Springfield  20:47:49  Wednesday  
8  Springfield  09:17:40     Friday  
9  Shelbyville  21:20:49  Wednesday

Información general del DataFrame:

In [3]:
#Obteniendo información general sobre los datos en df
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 65079 entries, 0 to 65078
Data columns (total 7 columns):
 #   Column    Non-Null Count  Dtype 
---  ------    --------------  ----- 
 0     userID  65079 non-null  object
 1   Track     63736 non-null  object
 2   artist    57512 non-null  object
 3   genre     63881 non-null  object
 4     City    65079 non-null  object
 5   time      65079 non-null  object
 6   Day       65079 non-null  object
dtypes: object(7)
memory usage: 3.5+ MB


In [4]:
df.describe()

userID  Track     artist  genre       City        time     Day
count      65079  63736      57512  63881        65079     65079   65079
unique     41748  39666      37806    268            2     20392       3
top     A8AE9169  Brand  Kartvelli    pop  Springfield  21:51:22  Friday
freq          76    136        136   8850        45360        14   23149

La tabla contiene siete columnas. Todas almacenan el mismo tipo de datos: objeto.

De acuerdo con la documentación:
- `'userID'` — identificador del usuario
- `'Track'` — título de la pista
- `'artist'` — nombre del artista
- `'genre'` — género
- `'City'` — ciudad del usuario
- `'time'` — el periodo de tiempo exacto en que se reprodujo la pista
- `'Day'` — día de la semana

Podemos ver tres problemas con el estilo en los nombres de las columnas:
1. Algunos nombres están en mayúsculas, otros en minúsculas.
2. Hay algunos espacios en algunos nombres.  
3. Los nombres de las columnas idealmente deben estar escritas en `snake_case`, por ejemplo, `userID` --> `user_id`.  

El número de valores de las columnas es diferente. Esto significa que los datos contienen valores ausentes.

### Conclusiones iniciales

Cada fila de la tabla almacena datos de la pista que fue reproducida. Algunas columnas describen la pista en sí: su título, el artista y el género. El resto transmite la información del usuario: la ciudad de la que viene, el tiempo que ha reproducido la pista. 

Está claro que los datos son suficientes para probar la hipótesis. Sin embargo, hay valores ausentes.

Para continuar, necesitamos preprocesar los datos.

## Etapa 2. Preprocesamiento de datos
Corrigimos el formato en los encabezados de las columnas y manejamos los valores ausentes. Además, comprobaremos si hay duplicados en los datos.

### Estilo del encabezado
Evaluamos los nombres de cada columna:

In [5]:
# la lista de los nombres de las columnas en la tabla df
print(df.columns)

Index(['  userID', 'Track', 'artist', 'genre', '  City  ', 'time', 'Day'], dtype='object')


Cambiamos los nombres de las columnas de acuerdo con las reglas del buen estilo:
* Si el nombre tiene varias palabras, utilizaremos snake_case
* Todos los caracteres deben ser minúsculas
* Eliminamos los espacios

In [6]:
# renombramos columnas
df = df.rename(columns={
    '  userID': 'user_id', 
    'Track': 'track',
    '  City  ': 'city',
    'time': 'time',
    'Day': 'day'
})

Comprueba el resultado. Imprime los nombres de las columnas una vez más:

In [7]:
# comprobando el nombre de las columnas
print(df.columns)

Index(['user_id', 'track', 'artist', 'genre', 'city', 'time', 'day'], dtype='object')


### Valores ausentes
Revisamos el número de valores ausentes en la tabla.

In [8]:
# calculando valores ausentes
df.isna().sum()

user_id       0
track      1343
artist     7567
genre      1198
city          0
time          0
day           0
dtype: int64

No todos los valores ausentes afectan a la investigación. Por ejemplo, los valores ausentes en la pista y artista no son cruciales. En este caso serán remplazamos con `unknown`.

En este caso, reemplazaremos los valores ausentes en `'track'`, `'artist'`, y `'genre'` con la string `'unknown'`.

In [9]:
columns_to_replace=['track','artist','genre']

for column in columns_to_replace:
    df[column].fillna('unknown', inplace=True) 

Nos aseguramos de que la tabla no contiene más valores ausentes.

In [10]:
# contando nuevamente valores ausentes
df.isna().sum()

user_id    0
track      0
artist     0
genre      0
city       0
time       0
day        0
dtype: int64

### Duplicados
Buscamos el número de duplicados obvios en la tabla:

In [11]:
# contando duplicado obvios
df.duplicated().sum()

3826

Existen **3826** duplicados. Utilizaré el método de `pandas` para deshacerme de los duplicados obvios:

In [12]:
# eliminando duplicados obvios
df=df.drop_duplicates().reset_index(drop = True)

Contamos los duplicados obvios una vez más para asegurarnos de que todos han sido eliminados:

In [13]:
# comprobando duplicados
df.duplicated().sum()

0

Ahora eliminaré los duplicados implícitos en la columna `genre` (si es que lo hay). Por ejemplo, el nombre de un género se puede escribir de varias formas. Dichos errores también pueden afectar a resultado.

In [14]:
# inspeccionando los nombres de géneros únicos
df.sort_values('genre',inplace=True) #inplace debe ponerse en True para que se actualice el DF

html_table = pd.DataFrame({'genre': df['genre'].unique()}).to_html(index=False)
display(HTML(html_table))

En la lista encontramos duplicados implícitos del género `hiphop`. Estos son nombres escritos incorrectamente o nombres alternativos para el mismo género.

En este caso, estos 3 géneros corresponden al genero `hiphop`:
* `hip`
* `hop`
* `hip-hop`

Para deshacernos de ellos facilmente, construiré la función `replace_wrong_genres()` con dos parámetros: 
* `wrong_genres=` — la lista de duplicados
* `correct_genre=` — la string con el valor correcto

La función coregirá los nombres en la columna `'genre'` de la tabla `df`, es decir, remplazará cada valor de la lista `wrong_genres` con el valor en `correct_genre`.

In [15]:
# función para reemplazar duplicados implícitos
def replace_wrong_genres(wrong_genres, correct_genre): 
    df['genre'] = df['genre'].replace(wrong_genres, correct_genre) 

Llamamos la función `replace_wrong_genres()` para retirar los duplicados implícitos (`hip`, `hop` y `hip-hop`) y los reemplace por `hiphop`:

In [16]:
# eliminando duplicados implícitos
duplicados = ['hip', 'hop','hip-hop'] # lista de nombres mal escritos
nombre_correcto = 'hiphop' # el nombre correcto
replace_wrong_genres(duplicados, nombre_correcto) 

Nos aseguramos que los nombres duplicados han sido eliminados. Imprimimos nuevamente la lista de valores únicos de la columna `'genre'`:

In [17]:
# revisando en busca de duplicados implícitos
#Ordenamos
df.sort_values('genre',inplace=True)
# convierte los valores únicos en una tabla HTML
html_table = pd.DataFrame({'genre': df['genre'].unique()}).to_html(index=False)
display(HTML(html_table))

### Conclusiones <a id='data_preprocessing_conclusions'></a>
Detectamos tres problemas con los datos:

- Estilos de encabezados incorrectos
- Valores ausentes
- Duplicados obvios e implícitos

Los encabezados han sido eliminados para conseguir que el procesamiento de la tabla sea más sencillo.

Todos los valores ausentes han sido reemplazados por `'unknown'`. Pero todavía tenemos que ver si los valores ausentes en `'genre'` afectan a nuestros cálculos.

La ausencia de duplicados hará que los resultados sean mas precisos y fáciles de entender.

Ahora ya podemos continuar probando las hipótesis. 

## Etapa 3. Prueba de hipótesis <a id='hypotheses'></a>

### Hipótesis 1: comparar el comportamiento del usuario en las dos ciudades <a id='activity'></a>

De acuerdo con la primera hipótesis, los usuarios de Springfield y Shelbyville escuchan música de forma distinta. Comprobaremos esto utilizando los datos de tres días de la semana: lunes, miércoles y viernes.

* Dividiré a los usuarios en grupos por ciudad.
* Veremos cuántas pistas reprodujo cada grupo el lunes, el miércoles y el viernes.

In [18]:
# contando las pistas reproducidas en cada ciudad
data_por_ciudad=df.groupby('city')
print(data_por_ciudad['track'].count())

city
Shelbyville    18512
Springfield    42741
Name: track, dtype: int64


Springfield ha reproducido más pistas que Shelbyville. Pero eso no implica necesariamente que los ciudadanos de Springfield escuchen música más a menudo. **Esta ciudad es simplemente más grande y hay más usuarios.**

Ahora agrupamos los datos por día de la semana y observamos el número de pistas reproducidas el lunes, miércoles y viernes.


In [19]:
# calculando las pistas reproducidas en cada uno de los tres días
data_por_dia=df.groupby('day')
print(data_por_dia['track'].count())

day
Friday       21840
Monday       21354
Wednesday    18059
Name: track, dtype: int64


El miércoles fue el día más silencioso de todos. Pero si consideramos las dos ciudades por separado podríamos llegar a una conclusión diferente.

Una vez que ya agrupamos por ciudad y día, ahora escribiré una función que agrupará ambos.

Crearé la función `number_tracks()` para calcular el número de canciones reproducidas en un determinado día y ciudad. Esta tendrá dos parámetros:
* día de la semana
* nombre de la ciudad

La función utilizará una variable para almacenar las filas de la tabla original, donde:
  * el valor de la columna `'day'` es igual al parámetro de día
  * el valor de la columna `'city'` es igual al parámetro de ciudad

Aplicaré un filtrado consecutivo con indexación lógica, filtrando primero por `day` y luego por `city`.

Después, contaré los valores de la columna `'user_id'` en la tabla resultante y regresaré este valor.

In [20]:
def number_tracks(day,city):
    #Se filtra por día
    track_list = df[df['day'] == day]
    
    #A la filtración anterior ahora se filtra por ciudad
    track_list = track_list[track_list['city'] == city]
    
    #Contamos
    track_list_count=track_list['user_id'].count()
    return track_list_count

Llamamos a `number_tracks()` seis veces, (3 días, 2 ciudades)

In [21]:
# el número de canciones reproducidas en Springfield el lunes
print(number_tracks('Monday','Springfield'))

15740


In [22]:
# el número de canciones reproducidas en Shelbyville el lunes
print(number_tracks('Monday','Shelbyville'))

5614


In [23]:
# el número de canciones reproducidas en Springfield el miércoles
print(number_tracks('Wednesday','Springfield'))

11056


In [24]:
# el número de canciones reproducidas en Shelbyville el miércoles
print(number_tracks('Wednesday','Shelbyville'))

7003


In [25]:
# el número de canciones reproducidas en Springfield el viernes
print(number_tracks('Friday','Springfield'))

15945


In [26]:
# el número de canciones reproducidas en Shelbyville el viernes
print(number_tracks('Friday','Shelbyville'))

5895


A continuación, crearé una nueva tabla con la siguiente informacion:
* Los nombres de las columnas son: `['city', 'monday', 'wednesday', 'friday']`
* Los datos son los 6 resultados conseguidos usando `number_tracks()`

In [27]:
entradas =['city', 'monday', 'wednesday', 'friday']
data_musica =[['Springfield',
               number_tracks('Monday','Springfield'),
               number_tracks('Wednesday','Springfield'),
               number_tracks('Friday','Springfield')],
              ['Shelbyville',
               number_tracks('Monday','Shelbyville'),
               number_tracks('Wednesday','Shelbyville'),
               number_tracks('Friday','Shelbyville')]]

tabla_dias_ciudades = pd.DataFrame(data=data_musica, columns=entradas)

In [28]:
# tabla con los resultados
tabla_dias_ciudades

city  monday  wednesday  friday
0  Springfield   15740      11056   15945
1  Shelbyville    5614       7003    5895

**Conclusiones**

Los datos revelan las diferencias en el comportamiento de los usuarios:

- En Springfield, el número de canciones reproducidas alcanzan el punto máximo los lunes y viernes mientras que los miércoles hay un descenso de la actividad.
- En Shelbyville, al contario, los usuarios escuchan más música los miércoles. La actividad de los usuarios los lunes y viernes es menor.

Así que la primera hipótesis parece ser correcta.

### Hipótesis 2: música al principio y al final de la semana <a id='week'></a>

De acuerdo con la segunda hipótesis, los lunes por la mañana y los viernes por la noche los ciudadanos de Springfield escuchan géneros que difieren de aquellos que los usuarios de Shelbyville disfrutan.

De la tabla principal, obtendré dos tablas, una para Springfield y otra para Shelbyville.
* Para Springfield — `spr_general`
* Para Shelbyville — `shel_general`

In [29]:
#Obteniendo la tabla Para Springfield (spr_general)
spr_general=df[df['city'] == 'Springfield']
spr_general.head(10)

user_id                       track                 artist     genre  \
31916  C40BC5DF            Shot in the Dark             Four80East      acid   
20264  B02ED3AB  Prelude – Ne Me Qui Te Pas                unknown  acoustic   
37784  77D3A7D9  Prelude – Ne Me Qui Te Pas                unknown  acoustic   
57580  23C64044           Don't Stop Me Now  Acoustic Heartstrings  acoustic   
26378  4D96EB78                 Eventuality          Jack Trammell    action   
9420   D3611A58                     Voyager         Jo Blankenburg    action   
12869  EED2E4E4    Rise from the Underworld                unknown    action   
4741   A51C478B                Baba O'Riley          Jack Trammell    action   
54927  60864740               Ball N' Chain      Big Mama Thornton     adult   
41456  F84E17DD             Come Back to Me    Sara Jackson-Holman     adult   

              city      time        day  
31916  Springfield  09:44:12     Friday  
20264  Springfield  13:31:03  Wednesday  
37784  Springfield  14:27:34     Monday  
57580  Springfield  13:30:01  Wednesday  
26378  Springfield  21:22:51     Monday  
9420   Springfield  14:20:29     Monday  
12869  Springfield  21:38:38     Monday  
4741   Springfield  14:36:09  Wednesday  
54927  Springfield  14:38:35  Wednesday  
41456  Springfield  13:59:23  Wednesday

In [30]:
# obteniendo la tabla para Shelbyville (shel_general)
shel_general=df[df['city'] == 'Shelbyville']
shel_general.head(10)

user_id                       track                artist     genre  \
3636   F10919ED          Ready For The Fire      Valley Of Wolves  acoustic   
14611  1C7D9E90  Prelude – Ne Me Qui Te Pas               unknown  acoustic   
13408   BF3BEA4                  Sweet Love           Anita Baker     adult   
1471   2ED60CF7               Ball N' Chain     Big Mama Thornton     adult   
47111  D939EE66                Perfect Rain          Curtis Walsh     adult   
9393   B24158ED                  Cellophane   Sara Jackson-Holman     adult   
38129  2029FD35          Dont Stop Believin          Rock Masters     adult   
18251  897AAB4E                Betcha Never                Cherie     adult   
47420  826C2C51                   Your Song             Ben Cocks     adult   
34021  4DCD818F            Blue Skies Again  Jessica Lea Mayfield     adult   

              city      time        day  
3636   Shelbyville  21:33:18     Monday  
14611  Shelbyville  21:47:42     Friday  
13408  Shelbyville  14:56:07     Monday  
1471   Shelbyville  21:55:27     Monday  
47111  Shelbyville  08:58:39     Monday  
9393   Shelbyville  14:18:42     Monday  
38129  Shelbyville  20:56:55  Wednesday  
18251  Shelbyville  09:30:39     Friday  
47420  Shelbyville  20:48:28     Friday  
34021  Shelbyville  09:05:48  Wednesday

Ahora escribiré la función genre_weekday() con cuatro parámetros de entrada:
* Una tabla para los datos (`df`)
* El día de la semana (`day`)
* Hora de inicio en formato 'hh:mm' (`time1`)
* Hora de terminoa en formato 'hh:mm' (`time2`)

La función devolverá información de los 15 géneros más populares de un día determinado en un período entre dos marcas de fecha y hora.

In [31]:
def genre_weekday(df,day,time1,time2):

    # Filtramos por día
    genre_df = df[df['day'] == day]
    
    # Filramos filas con valores de hora menores a la hora de termino
    genre_df = genre_df[genre_df['time'] < time2]

    # Filramos filas con valores de hora mayores a la hora de inicio
    genre_df = genre_df[genre_df['time'] > time1]
    
    # Agrupamos por género y luego contamos    
    genre_df_grouped = genre_df.groupby('genre')['genre'].count()
    #Agrupo por 'genre', selecciono la columna 'genre' y luego cuento

    # Ordenaremos el resultado en orden descendente (los géneros más populares aparecerán primero)
    genre_df_sorted = genre_df_grouped.sort_values(ascending=False) 
   
    # devolveremos la Serie con los 15 géneros más populares en un día y periodo de tiempo determinado
    return genre_df_sorted[:15]

Comparamos los resultados de la función `genre_weekday()`para Springfield y Shelbyville para
- Lunes por la mañana (de 7 a 11)
- Viernes por la tarde (de 17:00 a 23:00)

In [32]:
# llamando a la función para el lunes por la mañana en Springfield (utilizando la tabla de Springfield)
pd.DataFrame(genre_weekday(spr_general,'Monday','07:00','11:00'))

genre
genre             
pop            781
dance          549
electronic     480
rock           474
hiphop         286
ruspop         186
world          181
rusrap         175
alternative    164
unknown        161
classical      157
metal          120
jazz           100
folk            97
soundtrack      95

In [33]:
# llamando a la función para el lunes por la mañana en Shelbyville (utilizando la tabla de Shelbyville)
pd.DataFrame(genre_weekday(shel_general,'Monday','07:00','11:00'))

genre
genre             
pop            218
dance          182
rock           162
electronic     147
hiphop          80
ruspop          64
alternative     58
rusrap          55
jazz            44
classical       40
world           36
rap             32
soundtrack      31
rnb             27
metal           27

In [34]:
# llamando a la función para el viernes por la tarde en Springfield (utilizando la tabla de Springfield)
pd.DataFrame(genre_weekday(spr_general,'Friday','17:00','23:00'))

genre
genre             
pop            713
rock           517
dance          495
electronic     482
hiphop         273
world          208
ruspop         170
classical      163
alternative    163
rusrap         142
jazz           111
unknown        110
soundtrack     105
rnb             90
metal           88

In [35]:
# llamando a la función para el viernes por la tarde en Shelbyville (utilizando la tabla de Shelbyville)
pd.DataFrame(genre_weekday(shel_general,'Friday','17:00','23:00'))

genre
genre             
pop            256
rock           216
electronic     216
dance          210
hiphop          97
alternative     63
jazz            61
classical       60
rusrap          59
world           54
unknown         47
ruspop          47
soundtrack      40
metal           39
rap             36

**Conclusión**

Habiendo comparado los 15 géneros más populares del lunes por la mañana podemos concluir lo siguiente:

1. Los usuarios de Springfield y Shelbyville escuchan música similar. Los cinco géneros más populares son los mismos, solo rock y electrónica han intercambiado posiciones.

2. En Springfield el número de valores ausentes resultaron ser tan altos que el valor `'unknown'` llegó al décimo. Esto significa que los valores ausentes forman una parte considerable de los datos, lo que podría ser la base de la cuestión sobre la fiabilidad de nuestras conclusiones.

Para el viernes por la tarde, la situación es similar. Los géneros individuales varían algo pero, en general, los 15 más populares son parecidos en las dos ciudades.

De esta forma, la segunda hipótesis ha sido parcialmente demostrada:
* Los usuarios escuchan música similar al principio y al final de la semana.
* No hay una gran diferencia entre Springfield y Shelbyville. En ambas ciudades, el pop es el género más popular.

Sin embargo, el número de valores ausentes hace este resultado un tanto cuestionable. En Springfield, hay tantos que afectan a nuestros 15 más populares. De no faltarnos esos valores, las cosas podrían parecer diferentes.

### Hipótesis 3: preferencias de género en Springfield y Shelbyville <a id='genre'></a>

Hipótesis: Shelbyville ama la música rap. A los ciudadanos de Springfield les gusta más el pop.

Ahora agruparé la tabla `spr_general` por género y buscaré el número de canciones reproducidas de cada género con el método `count()`. Después, ordenaré el resultado en orden descendente y lo guardaré en `spr_genres`.

In [36]:
#Agrupo por 'genre'
#selecciono la columna 'genre'
#luego cuento (todo en una sola línea)
spr_genres=spr_general.groupby('genre')['genre'].count().sort_values(ascending=False)

Imprimimos las 10 primeras filas de `spr_genres`:

In [37]:
# imprimiendo las 10 primeras filas de spr_genres
spr_genres[:10]

genre
pop            5892
dance          4435
rock           3965
electronic     3786
hiphop         2096
classical      1616
world          1432
alternative    1379
ruspop         1372
rusrap         1161
Name: genre, dtype: int64

Ahora haré lo mismo con los datos de Shelbyville.

Agruparé la tabla `shel_general` por género y buscaré el número de canciones reproducidas de cada género. Después, ordenaré el resultado en orden descendente y lo guardaré en la tabla `shel_genres`:

In [38]:
#Agrupo por 'genre'
#selecciono la columna 'genre'
#luego cuento (todo en una sola línea)
shel_genres=shel_general.groupby('genre')['genre'].count().sort_values(ascending=False)

Imprimimos las 10 primeras filas de `shel_genres`:

In [39]:
# imprimiendo las 10 primeras filas de shel_genres
shel_genres[:10]

genre
pop            2431
dance          1932
rock           1879
electronic     1736
hiphop          960
alternative     649
classical       646
rusrap          564
ruspop          538
world           515
Name: genre, dtype: int64

**Conclusión**

La hipótesis ha sido parcialmente demostrada:
* La música pop es el género más popular en Springfield, tal como se esperaba.
* Sin embargo, la música pop ha resultado ser igual de popular en Springfield que en Shelbyville y el rap no estaba entre los 5 más populares en ninguna de las ciudades.


# Conclusiones <a id='end'></a>

Hemos probado las siguientes tres hipótesis:

1. La actividad de los usuarios difiere dependiendo del día de la semana y de las distintas ciudades. 
2. Los lunes por la mañana los residentes de Springfield y Shelbyville escuchan géneros distintos. Lo mismo ocurre con los viernes por la noche.
3. Los oyentes de Springfield y Shelbyville tienen distintas preferencias. En ambas ciudades, Springfield y Shelbyville, se prefiere el pop.

Tras analizar los datos, concluimos:

1. La actividad del usuario en Springfield y Shelbyville depende del día de la semana aunque las ciudades varían de diferentes formas. 

**La primera hipótesis ha sido aceptada completamente.**

2. Las preferencias musicales no varían significativamente en el transcurso de la semana en Springfield y Shelbyville. Podemos observar pequeñas diferencias en el orden los lunes, pero:
* En Springfield y Shelbyville la gente lo que más escucha es la música pop.

Así que **no podemos aceptar la segunda hipótesis**. También debemos tener en cuenta que el resultado podría haber sido diferente si no fuera por los valores ausentes.

3. Resulta que las preferencias musicales de los usuarios de Springfield y Shelbyville son bastante parecidas.

**La tercera hipótesis es rechazada**. Si hay alguna diferencia en las preferencias no se puede observar en los datos.